### PR TITLE
[MLIR] [TORCH] Add E2E support for `aten::narrow` op

### PR DESCRIPTION
**Pull request by JakopinA**
Currently still under development, tests pass with a test with 3 dimensional arrays but not 2 dimensional arrays. (Compiler crash)

(Now fixed!)

### CODE REVIEW

**Review by pashu123 — CHANGES_REQUESTED** [review]
Please clang format.

**Comment by pashu123** [comment]
Also, rebase to the latest main.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -223,6 +223,35 @@ class DecomposeAtenSelectIntOp : public OpRewritePattern<AtenSelectIntOp> {
...
+    Value length = op.length();
+
+    Value one =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(1));
+    
+    Value startPlusLength =
+        rewriter.create<AtenAddIntOp>(loc, one.getType(), start, length);
+
+    
+    // `aten.slice.tensor` doesn't squeeze the dim even when it's size 1 after
```
[the remark this answers is not in the document]
Comment (JakopinA) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:252]: Created some more comments, let me know if these help to make things more clear.

[the remark this answers is not in the document]
Comment (gpetters94) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:252]: The first two comments are probably unnecessary, it's pretty clear what the ConstantInt and AddInt ops are doing.

**Review by gpetters94 — APPROVED** [review]
LGTM

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (JakopinA) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:248]: Done

File: lib/Conversion/TorchToLinalg/DataMovement.cpp [hunk 2]
```diff
@@ -958,6 +958,7 @@ class ConvertAtenSliceTensorOp : public OpConversionPattern<AtenSliceTensorOp> {
   LogicalResult
   matchAndRewrite(AtenSliceTensorOp op, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
+
```
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/DataMovement.cpp:961]: The changes in this file are not needed

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 3]
```diff
@@ -228,6 +228,37 @@ class DecomposeAtenSelectIntOp : public OpRewritePattern<AtenSelectIntOp> {
...
+    Value dim = op.dim();
+    Value length = op.length();
+
+    Value one =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(1));
+    Value startPlusLength =
+        rewriter.create<AtenAddIntOp>(loc, one.getType(), start, length);
+
+    // `aten.slice.tensor` doesn't squeeze the dim even when it's size 1 after
+    // slicing, while `aten.select.int` does.
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:249]: This comment seems to be from the `DecomposeAtenSelectIntOp`. It should not be here

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 4]
```diff
@@ -228,6 +228,37 @@ class DecomposeAtenSelectIntOp : public OpRewritePattern<AtenSelectIntOp> {
...
+
+    // Replaces the AtenSliceTensorOp op, provides values from the nearly
+    // identical aten::narrow
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:252]: nit: This comment is not necessary. The code makes it very clear that we are replacing `AtenNarrowOp` with `AtenSliceTensorOp`

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 5]
```diff
@@ -228,6 +228,37 @@ class DecomposeAtenSelectIntOp : public OpRewritePattern<AtenSelectIntOp> {
...
+    rewriter.replaceOpWithNewOp<AtenSliceTensorOp>(
+        op, op.getResult().getType(), op.self(), /*dim*/ dim, /*start*/ start,
+        /*end*/ startPlusLength, /*step=*/one);
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:255]: the convention for adding the arg name as a comment is `/*argname=*/variable`. Some of the comments here are missing the `=` and there is an extra whitespace after the comments

**Review by ramiro050 — APPROVED** [review]
LGTM! I just have a few small change requests regarding style

File: python/torch_mlir_e2e_test/test_suite/slice_like.py [hunk 6]
```diff
@@ -366,3 +366,81 @@ def forward(self, x, src):
...
+        None,
+        ([-1, -1], torch.float32, True),
+    ])
+    def forward(self, x):
+        return torch.narrow(x, dim=1, start=0, length=2)
+
+
+@register_test_case(module_factory=lambda: NarrowVerticalTest2())
+def NarrowVerticalTest2_basic(module, tu: TestUtils):
+    module.forward(tu.rand(6,4))
```
Comment (vivekkhandelwal1) [inline python/torch_mlir_e2e_test/test_suite/slice_like.py:446]: Please add a new line.

**Review by vivekkhandelwal1 — APPROVED** [review]
LGTM

File: python/torch_mlir_e2e_test/test_suite/slice_like.py (hunk 6 above)
Comment (JakopinA) [reply python/torch_mlir_e2e_test/test_suite/slice_like.py:446]: What do you mean by this, where should the new line go?

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]
@JakopinA Please resolve the conflicts.

**Pull request merged**